ui/src/app/shared/components/cluster.tsx

```diff
@@ -0,0 +1,30 @@
+import {Tooltip} from 'argo-ui';
+import * as React from 'react';
+import {DataLoader} from '.';
+import * as models from '../models';
+import {services} from '../services';
+
+export const clusterName = (name: string | null) => {
+    return name || 'in-cluster';
+};
+
+export const clusterTitle = (cluster: models.Cluster) => {
+    return `${clusterName(cluster.name)} (${cluster.server})`;
+};
+
+const clusterHTML = (cluster: models.Cluster, showUrl: boolean) => {
+    const text = showUrl ? clusterTitle(cluster) : clusterName(cluster.name);
+    return <Tooltip content={cluster.server}><span>{text}</span></Tooltip>;
+};
+
+interface Props {
+    url: string;
+    showUrl?: boolean;
+}
+
+export const Cluster = React.memo((props: Props) => {
```

Comment:
we need to cache this look up, it's not extortionate, but could happen 100s of times if we're not careful

Reply:
this cache does not work -- yet

Reply:
I guess we need caching on app list page, right? I would load all clusters on list page and pass it via context. The Cluster can check if context has list of cluster and load cluster info via api if context is empty.

Reply:
yes - I don't know how to do this yet - React.memo does not do what I wanted

Reply:
working on PR to your fork